### DISCUSSION TITLE
`Cannot use 'with' and lambdas in the same function`

### DISCUSSION THREAD

**Discussion by ghost** [pascalabcnet/pascalabcnet#2945] [Q&A] [answered]
The question is, why? As you can see on https://github.com/pascalabcnet/pascalabcnet/discussions/2940, I didn't mix them. They are in the same function (the constructor), but they have nothing to do with each others. Why is there such a limitation?

**Comment by ghost** [comment]
@SunSerega `with` is very useful. Imagine with the code on https://github.com/pascalabcnet/pascalabcnet/discussions/2943 but you have to write `dialog.` that many times.

**Reply by ghost** [reply]
The limitation `Cannot use 'with' and lambdas in the same function` is too silly. Now I have to replace `with` to test your new code.

**Reply by SunSerega** [reply]
> you have to write `dialog.` that many times

That doesn't sound like a lot of effort.

And more so, if code in a single subprogram (function/procedure) becomes too complicated and big, accessing many things in one variable - consider making your own type, derived from whatever the type of this `dialog` is.
Then, you can write the initialization of the dialog's properties inside your new type's constructor. There you can reference properties directly, without `self.` before the name.

**Reply by SunSerega** [reply, marked as answer]
The limitation really is silly though. It's not that it's not possible, it's just a stub to not let you run into an unimplemented intersection of these 2 features. `with` as you use it is simple, but also not very useful. And the cases where it is useful - allow for very horrible interactions with the lambda capture mechanism, which will take a lot of time to get right even for a professional.

**Reply by ghost** [reply]
Please see the code on https://github.com/pascalabcnet/pascalabcnet/discussions/2943. Why should I make my own type? I created an object named `dialog` of type `TaskDialog` and use `with` to set the properties of it. What's wrong with that? Is repeating `dialog.` any better? I don't think so!

**Reply by SunSerega** [reply]
You don't have to create any types. Or even any subprograms. You can shove all the code into the main `begin-end.` and make it work. But it will be horrible.

Use the OOP. These libraries are designed around it.

**Reply by ghost** [reply]
Please help me rewrite this piece of code without using lambda. Thank you.

`HyperlinkClicked += (o,e)->begin end;`

**Reply by SunSerega** [reply]
> How?

You are already creating you own type `Form1`. Well, designer is creating it for you. But what I'm saying - you can go without it, just using
```
var f := new Form;
f.Prop1 := Val1;
// etc.
```
But that will only make it harder to write and read your code. Creating a simple type is far more trivial, than what designer is generating:
```
// bad name, but I have no idea what it's actually for
TaskDialogForShowTaskDialog = class(TaskDialog)
  constructor;
  begin
    // already within context, without the "with" operator
    WindowTitle := 'Task dialog sample';
    // etc.
  end;
end;
```
It's best to create a type like this for most things you are using and initing with many values.

**Reply by SunSerega** [reply]
> Please help me rewrite this piece of code without using lambda

That's also going to be far more code than what `with` allows you to save.

Well, anyway, to write it without lambda - you need to define the subprogram fully.
```
procedure LambdaReplacement(o: ???; e: ???);
begin
  
end;
...
HyperlinkClicked += LambdaReplacement;
```
Hover your mouse over `o` and `e` in code with lambda to find out what types you should write instead of `???`.
Alternatively, you can Ctrl+Click on `HyperlinkClicked` and explore its insides to find the `.Invoke` method, which would have the same signature as the subprogram you'll need to write to get rid of lambda.
